### PR TITLE
[GLUTEN-8108][VL] Correct the logic of null on failure behavior for try cast

### PR DESCRIPTION
**Pull request by acvictor**
## What changes were proposed in this pull request?
There are 3 failure behaviors defined for cast in Substrait
```
enum FailureBehavior { 
    FAILURE_BEHAVIOR_UNSPECIFIED = 0; 
    FAILURE_BEHAVIOR_RETURN_NULL = 1; 
    FAILURE_BEHAVIOR_THROW_EXCEPTION = 2; 
} 
```
Failure behavior should be set to 2 only when evaluation mode is ANSI. The current logic in Spark 3.4 and 3.5 uses the evaluation of `ansiEnabled = evalMode == EvalMode.ANSI || (evalMode == EvalMode.TRY && !canUseLegacyCastForTryCast)` to set the failure behavior and is incorrectly evaluated in some cases depending on the value of `canUseLegacyCastForTryCast`. `canUseLegacyCastForTryCast` should not matter in Velox/Gluten since it is used in Spark to optimize execution by avoiding adding a try-catch block around the cast.

A minimal repro shows that 
select **try_cast**(value as bigint) from tbl translates to Velox plan **cast**((trim( :VARCHAR, n0_0)) as BIGINT), implying that failure behavior is incorrectly set to 2 and so `SparkCastExpr` in Velox is created with `nullOnFailure_` set to false instead of true https://github.com/facebookincubator/velox/blob/main/velox/expression/ExprCompiler.cpp#L411.

(Fixes: \#ISSUE-ID https://github.com/apache/incubator-gluten/issues/8108)

## How was this patch tested?
Unit tests.

### CODE REVIEW

**Comment by acvictor** [comment]
@zhli1142015 @rui-mo can you please review?

File: gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala [hunk 1]
```diff
@@ -43,7 +43,10 @@ case class CastTransformer(substraitExprName: String, child: ExpressionTransform
   extends UnaryExpressionTransformer {
   override def doTransform(args: java.lang.Object): ExpressionNode = {
     val typeNode = ConverterUtils.getTypeNode(dataType, original.nullable)
-    ExpressionBuilder.makeCast(typeNode, child.doTransform(args), original.ansiEnabled)
+    ExpressionBuilder.makeCast(
+      typeNode,
+      child.doTransform(args),
+      original.evalMode == EvalMode.ANSI)
```
Comment (zhli1142015) [inline gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala:49]: This is added in spark 3.4. I think you only need to fix the behavior in spark 3.4 and after.

File: backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala [hunk 2]
```diff
@@ -1457,4 +1457,29 @@ abstract class ScalarFunctionsValidateSuite extends FunctionsValidateSuite {
         checkGlutenOperatorMatch[FilterExecTransformer](df)
     }
   }
+
+  test("Test cast and try_cast") {
```
Comment (zhli1142015) [inline backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala:1461]: Please add separate test case for each function

File: gluten-ut/spark35/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenCastWithAnsiOnSuite.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.GlutenTestsTrait
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.DataType
+
+class GlutenCastWithAnsiOnSuite extends CastWithAnsiOnSuite with GlutenTestsTrait {
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    SQLConf.get.setConf(SQLConf.ANSI_ENABLED, true)
```
Comment (zhli1142015) [inline gluten-ut/spark35/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenCastWithAnsiOnSuite.scala:26]: If ANSI is on, every thing should fallback, can you clarify what's this test suite for?

File: gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala (hunk 1 above)
Comment (zhli1142015) [reply gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala:49]: https://github.com/apache/incubator-gluten/blob/main/shims/spark35/src/main/scala/org/apache/gluten/sql/shims/spark35/Spark35Shims.scala#L485

File: backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala (hunk 2 above)
Comment (acvictor) [reply backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala]: Done.

File: gluten-ut/spark35/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenCastWithAnsiOnSuite.scala (hunk 3 above)
Comment (acvictor) [reply gluten-ut/spark35/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenCastWithAnsiOnSuite.scala]: I have removed this, thanks.

File: gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala (hunk 1 above)
Comment (acvictor) [reply gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala]: Thank you! Used this shim function.

File: gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala [hunk 4]
```diff
@@ -92,6 +92,16 @@ class VeloxTestSettings extends BackendTestSettings {
     .exclude(
       "Process Infinity, -Infinity, NaN in case insensitive manner" // +inf not supported in folly.
     )
+  enableSuite[GlutenTryCastSuite]
+    .exclude(
+      "Process Infinity, -Infinity, NaN in case insensitive manner" // +inf not supported in folly.
+    )
+    .exclude("ANSI mode: Throw exception on casting out-of-range value to byte type")
```
Comment (rui-mo) [inline gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: Would you add some comments on why these tests are excluded? I remember ANSI ON causes execution fallback, can the tests produce Spark expected result?

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala (hunk 4 above)
Comment (acvictor) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: These tests come from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastWithAnsiOnSuite.scala and when the suite is run separately with `SQLConf.get.setConf(SQLConf.ANSI_ENABLED, true)` falls back but when run as part of `GlutenTryCastSuite` where the conf is not set does not give expected results. (Also, the tests fail without the code changes in this PR.)

Comment (rui-mo) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: > the conf is not set does not give expected results

Thanks for providing the details. These tests are for ANSI OFF cases, right?

**Review by rui-mo — APPROVED** [review]

**Review by zhli1142015 — APPROVED** [review]
LGTM, thanks

**Comment by zhli1142015** [comment]
Run Gluten Clickhouse CI on x86

File: gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala [hunk 5]
```diff
@@ -43,7 +44,10 @@ case class CastTransformer(substraitExprName: String, child: ExpressionTransform
   extends UnaryExpressionTransformer {
   override def doTransform(args: java.lang.Object): ExpressionNode = {
     val typeNode = ConverterUtils.getTypeNode(dataType, original.nullable)
-    ExpressionBuilder.makeCast(typeNode, child.doTransform(args), original.ansiEnabled)
+    ExpressionBuilder.makeCast(
+      typeNode,
+      child.doTransform(args),
+      SparkShimLoader.getSparkShims.withAnsiEvalMode(original))
```
Comment (PHILO-HE) [inline gluten-substrait/src/main/scala/org/apache/gluten/expression/UnaryExpressionTransformer.scala:50]: Yes, I note Spark also sets `ansiEnabled = true` for `EvalMode.TRY`, which is for re-using the code logic of `EvalMode.ANSI`. The difference is, in TRY mode any exception is caught and then NULL is returned.
Velox has different implementation, which requires us to set this flag simply according to whether EvalMode is ANSI or not.

@acvictor, please leave some comments here to clarify, which should be helpful for future code maintenance.

https://github.com/apache/spark/blob/v3.5.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L471

File: gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala (hunk 4 above)
Comment (PHILO-HE) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: These tests are for ANSI ON.
@acvictor, for these tests in TryCastSuite, exceptions are not thrown, but null results are returned? It should be an expected behavior for Velox cast.
I cannot figure out how these exception checks can pass in vanilla Spark, assuming exception is handled internally in TRY mode.

Comment (acvictor) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: @PHILO-HE I don't fully understand what you mean. Are you saying Velox behaviour is incorrect?

Comment (PHILO-HE) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: @acvictor, please rebase the code. The core change looks good to me.

Comment (acvictor) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:99]: Updated, thank you!

**Review by PHILO-HE — APPROVED** [review]
Thanks!

**Pull request merged**